### PR TITLE
채팅 시간 오류

### PR DESCRIPTION
**Pull request by dlfmalr**
#26 채팅 자바스크립트 createDate 한국표준시간으로 수정

### CODE REVIEW

**Review by KYJ1108 — APPROVED** [review]

**Comment by dlfmalr** [comment]
<!doctype html>
<html lang="ko">

<head>
    <meta charset="UTF-8">
    <meta name="viewport"
          content="width=device-width, user-scalable=no,
initial-scale=1.0, maximum-scale=1.0, minimum-scale=1.0">
    <meta http-equiv="X-UA-Compatible" content="ie=edge">
    <link rel="stylesheet" th:href="@{|/styles.css|}">
    <link th:href="@***@***.***/dist/full.min.css|}"
rel="stylesheet"
          type="text/css" />
    <script src="https://cdn.tailwindcss.com"></script>
    <script src="https://kit.fontawesome.com/bccde956fc.js"
crossorigin="anonymous"></script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/stomp.js/2.3.3/stomp.min.js"></script>
    <script src="https://ajax.googleapis.com/ajax/libs/jquery/3.5.1/jquery.min.js"></script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/sockjs-client/1.5.0/sockjs.min.js"></script>
    <title>Chatting</title>
</head>

<body>
<div class="bg-gray-100 flex items-center justify-center h-screen">

    <div class="bg-white p-6 rounded-lg shadow-lg w-[800px] h-[800px]">
        <div class="flex flex-row items-center">
            <a href="/" style="display: flex; align-items: center;">
                나가기
            </a>
            <h1 class="text-2xl font-bold mb-4" style="margin:
0;"><span th:text="${opponentName}"></h1>
        </div>
        <div id="messages" class="w-[100%] h-[85%] overflow-y-auto mb-4">
            <div th:if="${!#lists.isEmpty(chatroom.chatMessageList)}">
                <div th:each="message : ${chatroom.chatMessageList}">
                    <div th:if="${message.sender.loginId != user.loginId}">
                        <div class="chat chat-start">
                            <div class="chat-image avatar">
                                <div class="w-10 rounded-full">
                                    <a
th:href="@{|/profile/${message.sender.loginId}|}">
                                        profile
                                    </a>
                                </div>
                            </div>
                            <div class="chat-header">
                                <div th:text="${message.sender.nickname}"></div>
                            </div>
                            <div class="chat-bubble">
                                <div th:text="${message.message}"></div>
                            </div>
                            <div class="chat-footer opacity-50">
                                <time class="text-xs opacity-50"
th:text="${#temporals.format(message.createDate, 'a hh:mm')}"></time>
                            </div>
                        </div>
                    </div>
                    <div th:if="${message.sender.loginId == user.loginId}">
                        <div class="chat chat-end">
                            <div class="chat-bubble">
                                <div th:text="${message.message}"></div>
                            </div>
                            <div class="chat-footer opacity-50">
                                <time class="text-xs opacity-50"
th:text="${#temporals.format(message.createDate, 'a hh:mm')}"></time>
                            </div>
                        </div>
                    </div>
                </div>
            </div>
            <div id="greetings">
                <div id="images"></div>
            </div>
        </div>
        <div class="flex flex-row items-center">
            <input id="message" type="text" class="border
border-gray-300 w-full rounded-lg p-2" style="height: 40px;" />
            <button id="sendButton" onclick="sendChat()"
style="height: 40px; display: flex; align-items: center;
justify-content: center; padding: 0 10px;">
                전송
            </button>
        </div>
    </div>
</div>
    <div>
        <script th:inline="javascript">
            function connect() {
                var socket = new SockJS("http://localhost:8080/ws-stomp");
                stompClient = Stomp.over(socket);
                const roomId = [[${ chatroom.id }]]
                console.log(roomId);
                stompClient.connect({}, function (frame) {
                    console.log('웹소켓 연결 성공:', frame);
                    //구독
                    stompClient.subscribe('/sub/talk/' + roomId,
function (chatMessageDto) {
                        showChat(JSON.parse(chatMessageDto.body));
                    });
                });
            }
            //html 에서 입력값, roomId 를 받아서 Controller 로 전달
            function sendChat() {
                const roomId = [[${ chatroom.id }]]
                const sender = [[${ user.loginId }]]
                if ($("#message").val() != "") {
                    stompClient.send("/pub/talk/" + roomId, {},
                        JSON.stringify({
                            'message': $("#message").val(),
                            'sender': sender
                        }));
                }
                $("#message").val('');
            }

            function showChat(chatMessageDto) {
                const sender = [[${ user.nickname }]];
                const formattedDate = formatTime(chatMessageDto.createDate);
                if (chatMessageDto.sender == sender) {
                    $("#greetings").append(
                        "<div class = 'chat chat-end  flex flex-col
items-end'>" +
                        "<div class = 'chat-bubble mr-2'>" +
chatMessageDto.message + "</div>" +
                        "<time class='text-xs opacity-50'>" +
formattedDate + "</time>" +
                        "</div>"
                    );
                } else {
                    $("#greetings").append(
                        "<div class='chat chat-start flex flex-row
items-center'>" +
                        "<div class='chat-image avatar'>" +
                        "<div class='w-10 rounded-full'>" +
                        "<a href='/profile/" + chatMessageDto.loginId + "'>" +
                        "profile" +
                        "</a>" +
                        "</div>" +
                        "</div>" +
                        "<div>" +
                        "<div class='chat-header'>" + chatMessageDto.sender +
                        "</div>" +
                        "<div class='chat-bubble'>" +
chatMessageDto.message + "</div>" +
                        "<div class='chat-footer opacity-50'>" +
                        "<time class='text-xs opacity-50'>" +
formattedDate + "</time>" +
                        "</div>" +
                        "</div>" +
                        "</div>"
                    );
                }
                var messageContainer = document.getElementById("messages");
                messageContainer.scrollTop = messageContainer.scrollHeight;
            }

            // 시간을 원하는 형식으로 포맷하는 함수
            function formatTime(rawDate) {
                const date = new Date(rawDate); // 문자열을 Date 객체로 변환

                // UTC 시간을 한국 표준시로 변환
                const offset = 9 * 60; // KST는 UTC+9 시간대이므로 9시간을 분으로 변환
                const kstDate = new Date(date.getTime() + offset *
60000); // 한국 시간으로 변환

                const hours = kstDate.getUTCHours();
                const minutes = kstDate.getUTCMinutes();
                const ampm = hours >= 12 ? '오후' : '오전';
                const formattedHours = hours % 12 || 12; // 0시를 12시로 변환
                const formattedMinutes = padNumber(minutes); // 분을 두 자리로 패딩

                return `${ampm} ${formattedHours}:${formattedMinutes}`;
            }

            // 숫자를 두 자리로 패딩하는 함수
            function padNumber(number) {
                return number.toString().padStart(2, '0');
            }



            $(function () {
                $("#message").keyup(function (event) {
                    if (event.keyCode === 13) {
                        sendChat();
                    }
                });
            });

            // 페이지 로드 후 실행되는 부분
            document.addEventListener('DOMContentLoaded', function () {
                connect();

                // 채팅창의 div 요소를 가져옵니다.
                var messageContainer = document.getElementById("messages");

                // 스크롤을 아래로 이동시킵니다.
                messageContainer.scrollTop = messageContainer.scrollHeight;
            });
        </script>
    </div>
</body>
</html>


2024년 6월 25일 (화) 오후 2:19, KYJ1108 ***@***.***>님이 작성:
